src/com/carolinarollergirls/scoreboard/event/ScoreBoardEventProviderImpl.java

```diff
@@ -0,0 +1,446 @@
+        synchronized(scoreBoardEventListeners) {
+            scoreBoardEventListeners.remove(listener);
+        }
+    }
+
+    public void unlink() { unlink(false); }
+    protected void unlink(boolean neighborsRemoved) {
+        for (Class<? extends Property> propertySet : properties) {
+            for (Property prop : propertySet.getEnumConstants()) {
+                if (prop instanceof AddRemoveProperty) {
+                    for (ValueWithId v : getAll((AddRemoveProperty) prop)) {
+                        ScoreBoardEventProviderImpl item = (ScoreBoardEventProviderImpl) v;
+                        if (item.getParent() == this) {
+                            item.unlink(neighborsRemoved || item instanceof NumberedScoreBoardEventProvider<?>);
+                        } else {
+                            remove((AddRemoveProperty) prop, item);
+                        }
+                    }
+                } else if (prop instanceof PermanentProperty &&
+                        ScoreBoardEventProvider.class.isAssignableFrom(prop.getType())) {
+                    set((PermanentProperty) prop, null);
+                }
+            }
+        }
+        for (ScoreBoardListener l : providers.keySet()) {
+            if (l instanceof UnlinkableScoreBoardListener) {
+                ((UnlinkableScoreBoardListener) l).unlink();
+            } else {
+                providers.get(l).removeScoreBoardListener(l);
+            }
+        }
+        getParent().remove(ownType, this);
+    }
+
+    public void addWriteProtection(Property prop) { addWriteProtectionOverride(prop, null); }
+    public void addWriteProtectionOverride(Property prop, Flag override) {
+        writeProtectionOverride.put(prop, override);
+    }
+    public boolean isWritable(Property prop, Flag flag) {
+        if (!writeProtectionOverride.containsKey(prop)) { return true; }
+        if (writeProtectionOverride.get(prop) == null) { return false; }
+        if (writeProtectionOverride.get(prop) == flag) { return true; }
+        return false;
+    }
+    
+    protected ScoreBoardListener setCopy(PermanentProperty targetProperty, ScoreBoardEventProvider sourceElement,
+            PermanentProperty sourceProperty, boolean readonly) {
+        ScoreBoardListener l = new ConditionalScoreBoardListener(sourceElement, sourceProperty,
+                new CopyScoreBoardListener(this, targetProperty));
+        sourceElement.addScoreBoardListener(l);
+        providers.put(l, sourceElement);
+        if (readonly) {
+            addWriteProtectionOverride(targetProperty, Flag.COPY);
+        } else {
+            reverseCopyListeners.put(targetProperty, 
+                    new CopyScoreBoardListener(sourceElement, sourceProperty, false));
+        }
+        set(targetProperty, sourceElement.get(sourceProperty), Flag.COPY);
+        return l;
+    }
+    protected ScoreBoardListener setCopy(final PermanentProperty targetProperty, ScoreBoardEventProvider indirectionElement,
```

Comment:
Comments would be useful to understand what indirection, inverse, copy, recalcuate etc. mean.

Reply:
It's still not clear to me what an indirectionProperty is, the comment is mostly repeating the type signature.